Add `FeatureModelConstraint` iterator and bindings

s9latimm opened the pull request:
- [x] Implementation
  - [x] Iterator (fixes ownership / access to unique pointers) 
- [x] Documentation
- [x] Tests
- [x] Python bindings

s9latimm committed:
Add iterator and bindings

M	bindings/python/vara-feature/pybind_FeatureModel.cpp
M	include/vara/Feature/FeatureModel.h
A	include/vara/Utils/UniqueIterator.h
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelParser.cpp

vulder reviewed (commented):
LGTM, add some tests and docs and we can merge this :)

s9latimm committed:
Merge remote-tracking branch 'origin/vara-dev' into f-FMConstraintBindings

A	.cmake-format
M	.github/workflows/build.yml
A	.github/workflows/pre-commit.yml
M	.gitmodules
A	.pre-commit-config.yaml
M	CMakeLists.txt
M	LICENSE.TXT
M	README.md
M	bindings/python/tests/CMakeLists.txt
M	bindings/python/vara-feature/CMakeLists.txt
A	bindings/python/vara-feature/pybind_FeatureModelBuilder.cpp
M	bindings/python/vara-feature/pybind_Init.cpp
A	cmake/FindZ3.cmake
M	cmake/vara_feature_macros.cmake
M	docs/CMakeLists.txt
M	docs/doxygen.cfg.in
M	external/csv-parser
A	external/z3
M	include/vara/CMakeLists.txt
M	include/vara/Feature/Feature.h
M	include/vara/Feature/FeatureModel.h
M	include/vara/Feature/FeatureModelBuilder.h
M	include/vara/Feature/StepFunction.h
M	include/vara/Feature/StepFunctionParser.h
M	include/vara/Sampling/SampleSetParser.h
M	include/vara/Sampling/SamplingMethods.h
A	include/vara/Solver/ConfigurationFactory.h
A	include/vara/Solver/Error.h
A	include/vara/Solver/Solver.h
A	include/vara/Solver/SolverFactory.h
A	include/vara/Solver/Z3Solver.h
M	lib/CMakeLists.txt
M	lib/Configuration/CMakeLists.txt
M	lib/Feature/CMakeLists.txt
M	lib/Sampling/CMakeLists.txt
M	lib/Sampling/SampleSetParser.cpp
A	lib/Solver/CMakeLists.txt
A	lib/Solver/SolverFactory.cpp
A	lib/Solver/Z3Solver.cpp
M	scripts/ci_utils/CMakeLists.txt
M	setup.py
M	tools/CMakeLists.txt
A	tools/config-generator/CMakeLists.txt
A	tools/config-generator/ConfigurationGenerator.cpp
M	tools/fm-viewer/CMakeLists.txt
M	tools/fm-viewer/FeatureModelViewer.cpp
M	unittests/CMakeLists.txt
M	unittests/Configuration/CMakeLists.txt
M	unittests/Feature/CMakeLists.txt
M	unittests/Feature/FeatureModelBuilder.cpp
M	unittests/Feature/FeatureModelParser.cpp
M	unittests/Feature/FeatureModelSxfm.cpp
M	unittests/Feature/FeatureModelWriter.cpp
M	unittests/Sampling/CMakeLists.txt
M	unittests/Sampling/SampleSetParserTests.cpp
M	unittests/Sampling/SampleSetWriterTests.cpp
A	unittests/Solver/BasicSolverTests.cpp
A	unittests/Solver/CMakeLists.txt
A	unittests/Solver/ConfigurationFactory.cpp
A	unittests/Solver/LargeCaseStudyTests.cpp
A	unittests/Solver/RealWorldCaseStudyTests.cpp
A	unittests/Solver/SolverFactory.cpp
A	unittests/Solver/Z3Tests.cpp
M	unittests/Utils/CMakeLists.txt
M	unittests/resources/CMakeLists.txt
M	unittests/resources/sxfm/sxfm_example.sxfm
M	unittests/resources/xml/error_mismatch_parent_child.xml
M	unittests/resources/xml/error_missing_child.xml
M	unittests/resources/xml/error_missing_exclude.xml
M	unittests/resources/xml/error_missing_implication.xml
M	unittests/resources/xml/error_missing_parent.xml
M	unittests/resources/xml/error_root_root.xml
A	unittests/resources/xml/test_dune_bin.xml
M	unittests/resources/xml/test_dune_num.xml
A	unittests/resources/xml/test_dune_num_explicit.xml
A	unittests/resources/xml/test_hipacc_bin.xml
A	unittests/resources/xml/test_hipacc_num.xml
A	unittests/resources/xml/test_hsqldb_num.xml
A	unittests/resources/xml/test_mixed_constraints.xml
M	unittests/resources/xml/vm.dtd

s9latimm committed:
Add tests and improve bindings

M	bindings/python/tests/TEST_INPUTS/example_feature_model.xml
M	bindings/python/tests/constraint/test_constraint.py
M	bindings/python/tests/feature_model/test_feature_model.py
M	bindings/python/vara-feature/pybind_Constraint.cpp
M	bindings/python/vara-feature/pybind_FeatureModel.cpp
M	include/vara/Feature/FeatureModel.h
M	unittests/resources/xml/vm.dtd

s9latimm committed:
Fix tidy

M	CMakeLists.txt
M	include/vara/Feature/FeatureModel.h

s9latimm committed:
Add comment

M	bindings/python/vara-feature/pybind_FeatureModel.cpp
